File: bindings/java/src/main/java/org/indunet/robot/bus/TopicPublisher.java
package org.indunet.robot.bus;

import com.sun.jna.Pointer;

/** Publisher bound to a topic on a {@link Node}. */
public final class TopicPublisher implements AutoCloseable {
    private Pointer ptr;

    TopicPublisher(Pointer ptr) {
        this.ptr = ptr;
    }

    public String topic() {
        return NativeUtils.takeCString(RobotBusC.Holder.INSTANCE.robot_bus_topic_publisher_topic(ptr));
    }

    public void publish(byte[] payload) {
        Errors.check(
                RobotBusC.Holder.INSTANCE.robot_bus_topic_publisher_publish(ptr, payload, payload.length),
                "publish");
    }

    @Override
    public void close() {
        RobotBusC.Holder.INSTANCE.robot_bus_topic_publisher_free(ptr);
        ptr = null;
    }
}

File: bindings/android/src/main/kotlin/org/indunet/robot/bus/TopicPublisher.kt
package org.indunet.robot.bus

import com.sun.jna.Pointer

/** Publisher bound to a topic on a [Node]. */
class TopicPublisher internal constructor(private var ptr: Pointer?) : AutoCloseable {
    fun topic(): String =
        NativeUtils.takeCString(RobotBusC.Holder.INSTANCE.robot_bus_topic_publisher_topic(ptr))

    fun publish(payload: ByteArray) {
        Errors.check(
            RobotBusC.Holder.INSTANCE.robot_bus_topic_publisher_publish(
                ptr,
                payload,
                payload.size.toLong(),
            ),
            "publish",
        )
    }

    override fun close() {
        RobotBusC.Holder.INSTANCE.robot_bus_topic_publisher_free(ptr)
        ptr = null
    }
}
